File: qqtech9-backend/src/models/runners.js
import { BusinessLogicError } from "../utils/errors.js";

export class Runners {
    constructor(runner){
        this.id = runner.id;
        this.ruleId = runner.rule_id ?? runner.ruleId;
        this.status = runner.status;
        this.lastRunAt = runner.last_run_at ?? runner.lastRunAt;
        this.createdAt = runner.created_at ?? runner.createdAt;
        this.updatedAt = runner.updated_at ?? runner.updatedAt;

        this.rule = runner.rule;
    }

    static fromArray(runnersArray) {   
        return runnersArray.map(runner => new Runners(runner));
    }
    
};

export class RunnerQueue{
    constructor(runnerQueue){
        this.id = runnerQueue.id;
        this.runnerId = runnerQueue.runner_id ?? runnerQueue.runnerId;
        this.status = runnerQueue.status;
        this.scheduledFor = runnerQueue.scheduled_for ?? runnerQueue.scheduledFor;
        this.queuedAt = runnerQueue.queued_at ?? runnerQueue.queuedAt;
        this.startedAt = runnerQueue.started_at ?? runnerQueue.startedAt;
        this.finishedAt = runnerQueue.finished_at ?? runnerQueue.finishedAt;
        this.attemptCount = runnerQueue.attempt_count ?? runnerQueue.attemptCount;

        this.rule = runnerQueue.rule;
    }

    static fromArray(runnerQueueArray) {
        return runnerQueueArray.map(runnerQueue => new RunnerQueue(runnerQueue));
    }
};

export class RunnerLogs {
    constructor(runnerLog){
        this.id = runnerLog.id;
        this.runnerId = runnerLog.runner_id ?? runnerLog.runnerId;
        this.queueId = runnerLog.queue_id ?? runnerLog.queueId;
        this.runTimeMs = runnerLog.run_time_ms ?? runnerLog.runTimeMs;
        this.executionStatus = runnerLog.execution_status ?? runnerLog.executionStatus;
        this.rowsAffected = runnerLog.rows_affected ?? runnerLog.rowsAffected;
        this.result = runnerLog.result;
        this.error = runnerLog.error;
        this.executedAt = runnerLog.executed_at ?? runnerLog.executedAt;

        this.rule = runnerLog.rule;
    }

    static fromArray(runnerLogsArray) {
        return runnerLogsArray.map(runnerLog => new RunnerLogs(runnerLog));
    }
};

File: python-workers/models/runners.py
from datetime import datetime
from sqlalchemy import UUID, Column, Integer, String, DateTime, Text, func, ForeignKey
from config.database import Base

class Runners(Base):
    __tablename__ = 'runners'

    id = Column(UUID, primary_key=True)
    rule_id = Column(UUID, ForeignKey('rules.id'))
    status = Column(String)
    last_run_at = Column(DateTime)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())


    rule = None

    @staticmethod
    def from_array(runners_array):
        return [Runners(**runner) if isinstance(runner, dict) else runner for runner in runners_array]

    def update_status(self, status):
        self.status = status
        return self

    def idle(self):
        self.status = 'IDLE'
        self.last_run_at = datetime.now()
        return self

    def fail(self):
        self.status = 'ERROR'
        self.last_run_at = datetime.now()
        return self


class RunnerQueue(Base):
    __tablename__ = 'runner_queue'

    id = Column(UUID, primary_key=True)
    runner_id = Column(UUID, ForeignKey('runners.id'))
    status = Column(String, default='PENDING')
    scheduled_for = Column(DateTime)
    started_at = Column(DateTime)
    finished_at = Column(DateTime)
    attempt_count = Column(Integer, default=0)

    # Campo virtual para rule
    rule = None

    def start(self):
        self.status = 'PROCESSING'
        self.started_at = datetime.now()
        return self

    def complete(self):
        self.status = 'COMPLETED'
        self.finished_at = datetime.now()
        return self

    def fail(self, attempt_count):
        self.status = 'FAILED'
        self.finished_at = datetime.now()
        self.attempt_count = attempt_count
        return self

    def retry(self, attempt_count):
        self.status = 'PENDING'
        self.attempt_count = attempt_count
        self.started_at = None
        self.finished_at = None
        return self

    @staticmethod
    def from_array(runner_queue_array):
        return [RunnerQueue(**rq) if isinstance(rq, dict) else rq for rq in runner_queue_array]


class RunnerLogs(Base):
    __tablename__ = 'runner_logs'

    id = Column(UUID, primary_key=True)
    runner_id = Column(UUID, ForeignKey('runners.id'))
    queue_id = Column(UUID, ForeignKey('runner_queue.id'))
    run_time_ms = Column(Integer)
    execution_status = Column(String)
    rows_affected = Column(Integer)
    result = Column(Text)
    error = Column(Text)
    executed_at = Column(DateTime, server_default=func.now())

    rule = None

    @staticmethod
    def from_array(runner_logs_array):
        return [RunnerLogs(**rl) if isinstance(rl, dict) else rl for rl in runner_logs_array]
